tests/test_client.py

```diff
@@ -1776,6 +1777,207 @@ def test_get_all_versions(client, user):
+
+
+def test_update_all_versions_guid_not_found(client, user):
+    bad_guid = "00000000-0000-0000-0000-000000000000"
+
+    update_data = {"acl": ["mock_acl"], "authz": ["mock_authz"]}
+    res = client.put(
+        "/index/{}/versions".format(bad_guid), json=update_data, headers=user
+    )
+    # Expect the operation to fail with 404 -- Guid not found
+    assert (
+        res.status_code == 404
+    ), "Expected update operation to fail with 404: {}".format(res.json)
+
+
+def test_update_all_versions_fail_on_bad_metadata(client, user):
+    """
+    When making an update request, endpoint should return 400 (User error) if
+    the metadata to update contains any fields that cannot be updated across all versions.
+    Currently the only allowed fields are ('acl', 'authz').
+    """
+    mock_acl_A = ["mock_acl_A1", "mock_acl_A2"]
+    mock_acl_B = ["mock_acl_B1", "mock_acl_B2"]
+    mock_authz_A = ["mock_authz_A1", "mock_authz_A2"]
+    mock_authz_B = ["mock_authz_B1", "mock_authz_B2"]
+
+    # SETUP
+    # -------
+    # create 1st version
+    data = get_doc(has_metadata=False, has_baseid=False)
+    data["acl"] = mock_acl_A
+    data["authz"] = mock_authz_A
+
+    res = client.post("/index/", json=data, headers=user)
+    assert res.status_code == 200
+    rec1 = res.json
+    assert rec1["did"]
+    baseid = rec1["baseid"]
+
+    # create 2nd version
+    res = client.post("/index/" + rec1["did"], json=data, headers=user)
+    assert res.status_code == 200
+    rec2 = res.json
+    assert rec2["baseid"] == baseid
+    # ----------
+
+    # Update all versions
+    update_data = {"urls": ["url_A"], "acl": mock_acl_B, "authz": mock_authz_B}
+    res = client.put(
+        "/index/{}/versions".format(baseid), json=update_data, headers=user
+    )
+    # Expect the operation to fail with 400
+    assert (
+        res.status_code == 400
+    ), "Expected update operation to fail with 400: {}".format(res.json)
+
+    # Expect all versions to retain the old acl/authz
+    res = client.get("/index/{}/versions".format(rec1["did"]))
+    assert res.status_code == 200, "Failed to get all versions"
+    for _, version in res.json.items():
+        assert version["acl"] == mock_acl_A
+        assert version["authz"] == mock_authz_A
+
+
+@pytest.mark.parametrize(
+    "mock_authz_permissions",
+    [
+        {
+            "test": {
+                "create": ["resource_A", "resource_B"],
+                "update": ["resource_A", "resource_B"],
+            },
+            "test_user": {"update": ["resource_B"]},
+        }
+    ],
+)
+def test_update_all_versions_fail_on_missing_permissions(client, user, use_mock_authz):
+    """
+    If user does not have the 'update' permission on any record, request should
+    fail with 403.
+    """
+    # SETUP
+    # -------
+    admin_user = user
+    test_user = {
+        "Authorization": (
+            "Basic " + base64.b64encode(b"test_user:test").decode("ascii")
```

Comment:
indexd can check authorization 2 ways:
- basic auth: if the username and password are correct, then the user is admin, can update anything
- token auth: uses a list of resources to ask arborist if the user has access

here you have 2 users that both use basic auth, so the resources they have access to in arborist should not matter - this test works because you mocked the whole `auth.authorize` function, but the logic is incorrect

you can remove the `use_mock_authz` fixture and just do one call with `user` fixture and one without, to make sure access is checked, or add more complex jwt/arborist mocking :/

Reply:
Gotchu, I think what I'll do if it makes sense is use a JWT/bearer token in the header and keep the mock auth.authorize function -- I wanted to test the specific behavior of 'user has access to one record but not both' which I couldn't figure out how to do with basic auth.